File: LeetCode_Marathons/LCM2/Hard/Maximum_Score_Words_Formed_by_Letters/a_bitmask.cpp
class Solution {
public:
    string getMask(int number, int maskSize) {
        string mask;
        while (number != 0) {
            mask = char( int('0') + (number % 2)) + mask;
            number /= 2;
        }
        int needToAppend = maskSize - mask.size();
        for (int jj = 0; jj < needToAppend; jj++) mask = "0" + mask;
        return mask;
    }

    int maxScoreWords(vector<string>& words, vector<char>& letters, vector<int>& score) {
        int n = words.size();
        int maskCount = pow(2, n);

        int ans = 0, scoreCounter = 0;
        map<char, int> mpLetters;

        for (auto myChr : letters) mpLetters[myChr]++;

        for (int maskInt = 1; maskInt < maskCount; maskInt++) {
            string mask = getMask(maskInt, n);
            map<char, int> mpLettersOfMask;

            for (int indexOfMask = 0; indexOfMask < n; indexOfMask++) {
                if (mask[indexOfMask] == '1') {
                    for (auto myChr : words[indexOfMask]) mpLettersOfMask[myChr]++;
                }
            }

            scoreCounter = 0;
            for (auto myPair : mpLettersOfMask) {
                char key = myPair.first;
                int value = myPair.second, value2 = mpLetters[key];
                if (value > value2) {
                    scoreCounter = 0;
                    break;
                } else {
                    scoreCounter += score[ int(key) - int('a') ] * value;
                }
            }

            if (scoreCounter > ans) ans = scoreCounter;
            mpLettersOfMask.clear();
        }
        
        return ans;
    }
};

File: LeetCode_Marathons/LCM2/Hard/Maximum_Score_Words_Formed_by_Letters/a_bitmask.java
import java.util.*;

public class Solution {
    public String getMask(int number, int maskSize) {
        StringBuilder mask = new StringBuilder();
        while (number != 0) {
            mask.insert(0, (char) ('0' + (number % 2)));
            number /= 2;
        }
        int needToAppend = maskSize - mask.length();
        for (int jj = 0; jj < needToAppend; jj++) {
            mask.insert(0, '0');
        }
        return mask.toString();
    }

    public int maxScoreWords(String[] words, char[] letters, int[] score) {
        int n = words.length;
        int maskCount = (int) Math.pow(2, n);

        int ans = 0, scoreCounter = 0;
        Map<Character, Integer> mpLetters = new HashMap<>();

        for (char myChr : letters) {
            mpLetters.put(myChr, mpLetters.getOrDefault(myChr, 0) + 1);
        }

        for (int maskInt = 1; maskInt < maskCount; maskInt++) {
            String mask = getMask(maskInt, n);
            Map<Character, Integer> mpLettersOfMask = new HashMap<>();

            for (int indexOfMask = 0; indexOfMask < n; indexOfMask++) {
                if (mask.charAt(indexOfMask) == '1') {
                    for (char myChr : words[indexOfMask].toCharArray()) {
                        mpLettersOfMask.put(myChr, mpLettersOfMask.getOrDefault(myChr, 0) + 1);
                    }
                }
            }

            scoreCounter = 0;
            for (Map.Entry<Character, Integer> entry : mpLettersOfMask.entrySet()) {
                char key = entry.getKey();
                int value = entry.getValue();
                int value2 = mpLetters.getOrDefault(key, 0);
                if (value > value2) {
                    scoreCounter = 0;
                    break;
                } else {
                    scoreCounter += score[key - 'a'] * value;
                }
            }

            if (scoreCounter > ans) {
                ans = scoreCounter;
            }
            mpLettersOfMask.clear();
        }

        return ans;
    }
}